text/0249-v2-experiments.md

```diff
@@ -373,6 +379,329 @@ Disadvantages:
+   be listed by autocomplete when a user types "grant".
+
+#### Alternatives
+
+Using `1` for visibility:
+
+- Rc1: for release candidate.
+- Alpha1.
+
+**How will it look in all supported languages:**
+
+> _to be completed once a naming scheme is decided_
+
+| scheme | TS  | Java | Python | .Net | Go  |
+| ------ | --- | ---- | ------ | ---- | --- |
+| PreV1  |     |      |        |      |     |
+
+### Discussion
+
+Advantages:
+
+- Experimental APIs can be used without declaring a fixed version on
+  `aws-cdk-lib`.
+- All APIs will be in `aws-cdk-lib`, preventing dependencies hell.
+- Since old versions of an API will only be deprecated and not removed, if
+  needed, we will be able to push critical updates to these versions as well.
+  For example, if we discover that `grantWrite` grants overly permissive
+  permissions, and the same occur in all of its experimental (deprecated)
+  predecessors, we will be able to push the fix to them as well. This will allow
+  us to get the fix to more customers in case of a critical fix.
+- Libraries will be able to use experimental APIs without locking their
+  consumers to a specific version of `aws-cdk-lib` (or any other module we
+  vend).
+- Same solution across all languages.
+- Allows finer granularity for opting-in to experimental APIs. Users who wants
+  to use a specific experimental API do not have to use either an experimental
+  version line, or an experimental module.
+
+Disadvantages:
+
+- Graduating a module will require a lot of code changes from our users to
+  remove the prefix/suffix.
+- User does not have a clear motivation to upgrade to a newer version.
+- Low adoption. Users might be hesitant to use an API with a  `PreX` in its name,
+  assuming it means "will break".
+- Cluttering the code base. Although most IDEs will mark deprecated properties
+  with a ~~strikethrough~~, they will still be listed by autocomplete.
+- Will force some pretty long and ugly names on our APIs. Many previews APIs
+  will result in a less aesthetic user code.
+- A lot of deprecated code in aws-cdk-lib, possibly blowing up `aws-cdk-lib`.
+  This might not be a real concern as we can reuse a lot the code between
+  different version of an API.
+
+**How can we encourage users to upgrade to a newer version?**
+
+When a new minor version introduces a breaking change to an API, users have a
+clear motivation to upgrade to the new version of the API - they must do so in
+order to upgrade to a newer version of the AWS CDK. If the API is only
+deprecated, users have no motivation to upgrade to a newer version of the API,
+even worse, they might not be aware that a newer version exists. One way to
+encourage users to upgrade to a newer version of an API, is to supply tools that
+will inform users. For example, we can add a capability to the `cdk doctor`
+command that will notify users that their CDK application is using deprecated
+experimental APIs.
+
+Executing `cdk doctor` will print:
```

Comment:
Good idea but probably not `cdk doctor`. This is run only when there's a problem, but often not even then.
Maybe as part of `synth` or `deploy` itself.